test/integration/http_subset_lb_integration_test.cc

```diff
@@ -216,4 +216,35 @@ TEST_P(HttpSubsetLbIntegrationTest, SubsetLoadBalancer) {
   runTest(type_b_request_headers_, "b");
 }
 
+// Tests subset-compatible load balancer policy without metadata does not crash on initialization.
+TEST_P(HttpSubsetLbIntegrationTest, SubsetLoadBalancerNoMetadata) {
+  config_helper_.addConfigModifier([&](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* static_resources = bootstrap.mutable_static_resources();
+    auto* cluster = static_resources->mutable_clusters(0);
+
+    // Set single_host_per_subset to be true
+    auto* subset_selector = cluster->mutable_lb_subset_config()->mutable_subset_selectors(0);
+    subset_selector->set_single_host_per_subset(true);
+
+    // Clear the metadata for each host
+    auto* load_assignment = cluster->mutable_load_assignment();
+    auto* endpoints = load_assignment->mutable_endpoints(0);
+    for (uint32_t i = 0; i < num_hosts_; i++) {
+      auto* lb_endpoint = endpoints->mutable_lb_endpoints(i);
+      lb_endpoint->clear_metadata();
+    }
+  });
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto response = codec_client_->makeHeaderOnlyRequest(
+      Http::TestRequestHeaderMapImpl{{":method", "GET"},
+                                     {":path", "/test/long/url"},
+                                     {":scheme", "http"},
+                                     {":authority", "host"}});
+  response->waitForEndStream();
+  EXPECT_TRUE(response->complete());
+  EXPECT_THAT(response->headers(), Http::HttpStatusIs("404"));
```

Comment:
Do you know why this is giving a 404 instead of 503?
It might be configurable, but the default is 503 https://github.com/envoyproxy/envoy/blob/7adc0393221467fcae18052c1f7caf0cb6048d2e/api/envoy/config/route/v3/route_components.proto#L564

Reply:
The configuration used to test the issue is below:  cluster and host are configured.  It looks like in this case, 404 is returned.

static_resources:
  clusters:
    - name: cluster_0
      connect_timeout: 5s
      lb_subset_config:
        subset_selectors:
          - keys:
              - type
            single_host_per_subset: true
      load_assignment:
        cluster_name: cluster_0
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 46423
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 40245
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 39563
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 37041

Reply:
The configuration for this test is below: cluster and endpoints are configured. It looks like with this configuration, 404 is returned.

static_resources:
  clusters:
    - name: cluster_0
      connect_timeout: 5s
      lb_subset_config:
        subset_selectors:
          - keys:
              - type
            single_host_per_subset: true
      load_assignment:
        cluster_name: cluster_0
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 46423
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 40245
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 39563
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 37041

Reply:
Actually now that I think about it I think we're probably hitting a NR here and not actually hitting the bad code. I think the test file sets up a requirement for the header `x-type` to be set to either `a` or `b` in order to actually route to the upstream. Can you take a look and make sure we're actually attempting to route via the subset lb?

Reply:
Thanks! Let me take a look.

Reply:
1) The bad code in subset_lb.cc:SubsetLoadBalancer::rebuildSingle() is hit by configuration update, not by traffic. Which means in the newly added test code:
 TEST_P(HttpSubsetLbIntegrationTest, SubsetLoadBalancerSingleHostPerSubsetNoMetadata) 

 initialize(), which eventually call SubsetLoadBalancer::rebuildSingle() can trigger crash. We reproduced with this, also verified the fix with it. 

2) The original test function: TEST_P(HttpSubsetLbIntegrationTest, SubsetLoadBalancer) setup the config with metadata as 
...
        - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 34033
                metadata:
                  filter_metadata:
                    envoy.lb:
                      type: a
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 35437
                metadata:
                  filter_metadata:
                    envoy.lb:
                      type: b

...
 also call runTest() with type_a and type_b as in TEST_P(..., SubsetLoadBalancer), in this test, Envoy is expected to route via subset lb with this test.  

However,   the newly added test function doesn't have type a/b metadata config, neither call runTest with the two different request types,  it probably won't route via the subset lb.

3) In the newly added test function, the code after initialize(), i.e, creating HTTP connection and sending a HTTP request and verify the response is not specially testing the code change in SubsetLoadBalancer::rebuildSingle(), but a general purpose test, i.e, just to send some traffic, and verify response as expected. 

Please let me know whether this addressed your question.

Thanks!